src/navmesh.js

```diff
@@ -373,98 +375,102 @@ export class NavMesh {
         return path.reverse();
     }
 
+    /** Funnel algorithm, following https://medium.com/@reza.teshnizi/the-funnel-algorithm-explained-visually-41e374172d2d */
     _funnel(from, to, path) {
         if (path.length === 0) {
             throw new Error("Path cannot be empty.");
         } else if (path.length === 1) {
             return [from, to];
         }
 
-        const points = [from];
-        let edge1 = null;
-        let edge2 = null;
+        const tail = [from];
+        const left = [];
+        const right = [];
 
-        for (let i = 0; i < path.length - 1; i++) {
+        // Initialize funnel
+        const initialPortal = path[0].neighbors[path[1]._uuid].portal;
+        left.push(initialPortal.p1);
+        right.push(initialPortal.p2);
+
+        // Iterate over portals
+        for (let i = 1; i < path.length - 1; i++) {
             const poly = path[i];
-            const next = path[i + 1];
-            const portal = poly.neighbors[next._uuid].portal;
-
-            // Calculate portal edges
-            let [newEdge1, newEdge2] = this._pointPortalEdges(from, portal);
-
-            if (edge1 !== null && edge2 !== null) {
-                // Shrink funnel edges
-                newEdge1 = this._funnelEdge(edge1, edge2, newEdge1);
-                newEdge2 = this._funnelEdge(edge2, edge1, newEdge2);
-
-                if (newEdge1 === null || newEdge2 === null) {
-                    const edge = newEdge1 === null ? edge2 : edge1;
-                    const newPath = this._splitAt(
-                        // We only need to check the polygon up to now, not the future
-                        // ones. This can have serious performance implications,
-                        // as the function is recursive.
-                        path.slice(i - 1),
-                        edge.p2
-                    );
-                    points.push(...this._funnel(edge.p2, to, newPath));
-                    break;
-                }
-            }
+            const nextPoly = path[i + 1];
+            const portal = poly.neighbors[nextPoly._uuid].portal;
+            // The portal end points are in clockwise order, viewed from the inside of the polygon.
 
-            edge1 = newEdge1;
-            edge2 = newEdge2;
-
-            // If we are at the end of the path,
-            // just add the last edge, which is from -> to
-            if (i === path.length - 2) {
-                const toEdge = new Edge(from, to);
-                let newToEdge = this._funnelEdge(edge1, edge2, toEdge);
-                if (newToEdge === edge1) {
-                    points.push(edge1.p2);
-                } else if (newToEdge === null) {
-                    points.push(edge2.p2);
-                }
+            // Extend funnel on the left
+            this._extendFunnel(tail, left, right, true, portal.p1);
+
+            // Extend funnel on the right
+            this._extendFunnel(tail, left, right, false, portal.p2);
```

Comment:
Since we're doing the same thing twice, I would collect the two calls to `_extendFunnel` in a separate function.